Essentials/src/main/java/com/earth2me/essentials/User.java

```diff
@@ -322,44 +333,94 @@ public void setLogoutLocation() {
 
     @Override
     public void requestTeleport(final User player, final boolean here) {
-        teleportRequestTime = System.currentTimeMillis();
-        teleportRequester = player == null ? null : player.getBase().getUniqueId();
-        teleportRequestHere = here;
-        if (player == null) {
-            teleportLocation = null;
-        } else {
-            teleportLocation = here ? player.getLocation() : this.getLocation();
+        final TpaRequest token = teleportRequestQueue.getOrDefault(player.getName(), new TpaRequest(player.getName(), player.getUUID()));
+        token.setTime(System.currentTimeMillis());
+        token.setHere(here);
+        token.setLocation(here ? player.getLocation() : this.getLocation());
+
+        // Handle max queue size
+        teleportRequestQueue.remove(token.getName());
+        if (teleportRequestQueue.size() >= ess.getSettings().getTpaMaxAmount()) {
+            String lastKey = null;
+            for (Map.Entry<String, TpaRequest> entry : teleportRequestQueue.entrySet()) {
+                lastKey = entry.getKey();
+            }
+            teleportRequestQueue.remove(lastKey);
         }
+
+        // Add request to queue
+        teleportRequestQueue.put(token.getName(), token);
     }
 
     @Override
+    @Deprecated
     public boolean hasOutstandingTeleportRequest() {
-        if (getTeleportRequest() != null) { // Player has outstanding teleport request.
-            final long timeout = ess.getSettings().getTpaAcceptCancellation();
-            if (timeout != 0) {
-                if ((System.currentTimeMillis() - getTeleportRequestTime()) / 1000 <= timeout) { // Player has outstanding request
-                    return true;
-                } else { // outstanding request expired.
-                    requestTeleport(null, false);
-                    return false;
-                }
-            } else { // outstanding request does not expire
-                return true;
-            }
-        }
-        return false;
+        return getNextTpaToken(false, false, false) != null;
     }
 
-    public UUID getTeleportRequest() {
-        return teleportRequester;
+    public Collection<String> getPendingTpaKeys() {
+        return teleportRequestQueue.keySet();
     }
 
-    public boolean isTpRequestHere() {
-        return teleportRequestHere;
+    @Override
+    public boolean hasPendingTpaRequests(boolean inform, boolean excludeHere) {
+        return getNextTpaToken(inform, false, excludeHere) != null;
+    }
+
+    public boolean hasOutstandingTpaRequest(String playerUsername, boolean here) {
+        final TpaRequest token = getOutstandingTpaRequest(playerUsername, false);
+        return token != null && token.isHere() == here;
+    }
+
+    public TpaRequest getOutstandingTpaRequest(String playerUsername, boolean inform) {
```

Comment:
```suggestion
    public @Nullable TpaRequest getOutstandingTpaRequest(String playerUsername, boolean inform) {
```

Applies to the rest of nullable stuff